rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+        - a single constraint. The constraint can be a Spec object, a type, or None.
+
+    Args:
+        spec: The spec to convert to canonical format.
+
+    Returns:
+        The canonical format of the spec.
+    """
+    # convert spec of form list of nested_keys to model_spec with None leaves
+    if isinstance(spec, list):
+        spec = [(k,) if isinstance(k, str) else k for k in spec]
+        return SpecDict({k: None for k in spec})
+
+    # convert spec of form tree of constraints to model_spec
+    if isinstance(spec, abc.Mapping):
+        spec = SpecDict(spec)
+        for key in spec:
+            # if values are types or tuple of types, convert to TypeSpec
+            if isinstance(spec[key], (type, tuple)):
+                spec[key] = TypeSpec(spec[key])
+        return spec
+
+    if isinstance(spec, type):
+        return TypeSpec(spec)
+
+    # otherwise, assume spec is already in canonical format
+    return spec
+
+
+def _should_validate(
+    cls_instance: object, method: Callable, tag: str = "input"
+) -> bool:
+    """Returns True if the spec should be validated, False otherwise.
+
+    The spec should be validated if the method is not cached (i.e. there is no cache
+    storage attribute in the instance) or if the method is already cached. (i.e. it
+    exists in the cache storage attribute)
+
+    Args:
+        cls_instance: The class instance that the method belongs to.
+        method: The method to apply the spec checking to.
+        tag: The tag of the spec to check. Either "input" or "output". This is used
+        internally to defined an internal cache storage attribute based on the tag.
+
+    Returns:
+        True if the spec should be validated, False otherwise.
+    """
+    cache_store = getattr(cls_instance, f"__checked_{tag}_specs_cache__", None)
+    return cache_store is None or method.__name__ not in cache_store
+
+
+def _validate(
+    *,
+    cls_instance: object,
+    method: Callable,
+    data: Mapping[str, Any],
+    spec: Spec,
+    filter: bool = False,
+    tag: str = "input",
+) -> NestedDict:
+    """Validate the data against the spec.
+
+    Args:
+        cls_instance: The class instance that the method belongs to.
+        method: The method to apply the spec checking to.
+        data: The data to validate.
+        spec: The spec to validate against.
+        filter: If True, the data will be filtered to only include the keys that are
+            specified in the spec.
+        tag: The tag of the spec to check. Either "input" or "output". This is used
+            internally to defined an internal cache storage attribute based on the tag.
+
+    Returns:
+        The data, filtered if filter is True.
+    """
+    is_mapping = isinstance(spec, SpecDict)
+    cache_miss = _should_validate(cls_instance, method, tag=tag)
+
+    if is_mapping:
+        if not isinstance(data, Mapping):
+            raise ValueError(f"{tag} must be a Mapping, got {type(data).__name__}")
```

Comment:
nit: `is_mapping` is created and used only once. I don't think it adds readability like this. It would be clearer to me if it was inlined.

Reply:
I just figured I should use abc.Mapping here as well. Thanks for the feedback :)